google/cloud/forseti/scanner/audit/enabled_apis_rules_engine.py

```diff
@@ -0,0 +1,401 @@
+        """
+        for (i, rule) in enumerate(rule_defs.get('rules', [])):
+            self.add_rule(rule, i)
+
+    def add_rule(self, rule_def, rule_index):
+        """Add a rule to the rule book.
+
+        The rule supplied to this method is the dictionary parsed from
+        the rules definition file.
+
+        For example, this rule...
+
+            # rules yaml:
+            rules:
+              - name: a rule
+                mode: whitelist
+                resource:
+                  - type: project
+                    resource_ids:
+                      - my-project-123
+                services:
+                  - 'compute.googleapis.com'
+                  - 'storage-component.googleapis.com'
+                  - 'storage-api.googleapis.com'
+
+        ... gets parsed into:
+
+            {
+                'name': 'a rule',
+                'mode': 'whitelist',
+                'resource': {
+                    'type': 'project',
+                    'resource_ids': ['my-project-id']
+                },
+                'services': [
+                    'compute.googleapis.com',
+                    'storage-component.googleapis.com',
+                    'storage-api.googleapis.com'
+                ]
+            }
+
+        Args:
+            rule_def (dict): Contains rule definition properties.
+            rule_index (int): The index of the rule from the rule definitions.
+                Assigned automatically when the rule book is built.
+        """
+        self._rules_sema.acquire()
+
+        try:
+            resources = rule_def.get('resource')
+            mode = rule_def.get('mode')
+            services = rule_def.get('services')
+            if not resources or not services or mode not in _RULE_MODES:
+                raise audit_errors.InvalidRulesSchemaError(
+                    'Faulty rule {}'.format(rule_index))
+
+            for resource in resources:
+                resource_ids = resource.get('resource_ids')
+                resource_type = None
+                try:
+                    resource_type = resource_mod.ResourceType.verify(
+                        resource.get('type'))
+                except resource_errors.InvalidResourceTypeError:
+                    raise audit_errors.InvalidRulesSchemaError(
+                        'Missing resource type in rule {}'.format(rule_index))
+
+                if not resource_ids or len(resource_ids) < 1:
+                    raise audit_errors.InvalidRulesSchemaError(
+                        'Missing resource ids in rule {}'.format(rule_index))
+
+                # For each resource id associated with the rule, create a
+                # mapping of resource => rules.
+                for resource_id in resource_ids:
+                    if resource_id == '*':
+                        # Wild-card rules apply to all projects.
+                        gcp_resource = resource_util.create_resource(
```

Comment:
Is it possible for a user to use wild card matching and have the `resource_type` to be `folder` like the following:
 resource:
      - type: folder
        resource_ids:
         - '*'

Reply:
I decided to change the rule spec to only allow the wild-card rule to be specified with type=project. There is no need to have 3 different ways to specify "all projects", and there was some ambiguity on whether a wild-card folder rule applies to all projects, or only projects that are under a folder.